numpy/ma/core.py

```diff
@@ -2980,7 +2980,8 @@ def __array_finalize__(self, obj):
             # side-effecting 'obj' as well.
             if (obj.__array_interface__["data"][0]
                     != self.__array_interface__["data"][0]):
-                _mask = _mask.copy()
+                _mask = _mask.copy(order="F" if self.flags.f_contiguous
+                                   else "C")
```

Comment:
Does this keep the old behaviour on `x.astype(bool)`?

Reply:
What old behaviour are you thinking off? In the old `astype`, the mask order was always `C` - I changed this only to match your #9347 (and thus pass your tests).

Reply:
I'm thinking about how this has changed from `_mask.astype` to `_mask.copy` - what happens if the mask changes to a different structured type?

Reply:
Ah, I see: maybe I should also address the "This looks like a bug" at line 2967... (just expand above this change).

Reply:
So, this would be a test:
```
a = np.ma.MaskedArray([1, 2], mask=[True, False])
a.astype('i4,i4')
masked_array(data = [(--, --) (2, 2)],
             mask = [( True,  True) (False, False)],
       fill_value = (999999, 999999),
            dtype = [('f0', '<i4'), ('f1', '<i4')])
```
(note that `view` on this dtype fails in master (and also in numpy 1.13, quite possibly much longer...)